client/asset/btc/spv.go

```diff
@@ -0,0 +1,1631 @@
+	}
+
+	if !matchFound {
+		return
+	}
+
+	// There is at least one match. Pull the block.
+	block, err := w.cl.GetBlock(blockHash)
+	if err != nil {
+		w.log.Errorf("neutrino GetBlock error: %v", err)
+		return
+	}
+
+	for _, msgTx := range block.MsgBlock().Transactions {
+		newlyDiscovered := findRedemptionsInTx(ctx, true, reqs, msgTx, w.chainParams)
+		for outPt, res := range newlyDiscovered {
+			discovered[outPt] = res
+		}
+	}
+	return
+}
+
+// findRedemptionsInMempool is unsupported for SPV.
+func (w *spvWallet) findRedemptionsInMempool(ctx context.Context, reqs map[outPoint]*findRedemptionReq) (discovered map[outPoint]*findRedemptionResult) {
+	return
+}
+
+// confirmations looks for the confirmation count and spend status on a
+// transaction output that pays to this wallet.
+func (w *spvWallet) confirmations(txHash *chainhash.Hash, vout uint32) (blockHash *chainhash.Hash, confs uint32, spent bool, err error) {
+	details, err := w.wallet.walletTransaction(txHash)
+	if err != nil {
+		return nil, 0, false, err
+	}
+	if details == nil {
+		return nil, 0, false, WalletTransactionNotFound
+	}
+
+	if details.Block.Hash != (chainhash.Hash{}) {
+		blockHash = &details.Block.Hash
+	}
+
+	for _, credit := range details.Credits {
+		if credit.Index == vout {
+			syncBlock := w.wallet.syncedTo() // Better than chainClient.GetBestBlockHeight() ?
+			return blockHash, uint32(confirms(details.Block.Height, syncBlock.Height)), credit.Spent, nil
+		}
+	}
+
+	return blockHash, 0, false, SpentStatusUnknown
+}
+
+// getTransaction retrieves the specified wallet-related transaction.
+// This is pretty much a copy-past from btcwallet 'gettransaction' JSON-RPC
+// handler.
+func (w *spvWallet) getTransaction(txHash *chainhash.Hash) (*GetTransactionResult, error) {
+	// Option # 1 just copies from UnstableAPI.TxDetails. Duplicating the
+	// unexported bucket key feels dirty.
+	//
+	// var details *wtxmgr.TxDetails
+	// err := walletdb.View(w.Database(), func(dbtx walletdb.ReadTx) error {
+	// 	txKey := []byte("wtxmgr")
+	// 	txmgrNs := dbtx.ReadBucket(txKey)
+	// 	var err error
+	// 	details, err = w.TxStore.TxDetails(txmgrNs, txHash)
+	// 	return err
+	// })
+
+	// Option #2
+	// This is what the JSON-RPC does (and has since at least May 2018).
+	details, err := w.wallet.walletTransaction(txHash)
+	if err != nil {
+		return nil, err
+	}
+	if details == nil {
+		return nil, WalletTransactionNotFound
+	}
+
+	syncBlock := w.wallet.syncedTo()
+
+	// TODO: The serialized transaction is already in the DB, so
+	// reserializing can be avoided here.
+	var txBuf bytes.Buffer
+	txBuf.Grow(details.MsgTx.SerializeSize())
+	err = details.MsgTx.Serialize(&txBuf)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &GetTransactionResult{
+		TxID:         txHash.String(),
+		Hex:          txBuf.Bytes(), // 'Hex' field name is a lie, kinda
+		Time:         uint64(details.Received.Unix()),
+		TimeReceived: uint64(details.Received.Unix()),
+	}
+
+	if details.Block.Height != -1 {
```

Comment:
For sure `details.Block` is never nil? Also `details.MsgTx` above. Assuming they are pointers.

Reply:
Not pointers I see. Ty.